### PR TITLE
Add lang specific font stacks for CJK

### PR DESCRIPTION
**Pull request by zeripath**
The PR #5983 proposed deleting the YaHei font from the default font stack in order to fix issues with the presentation of Kanji in Japanese being forced in to simplified Chinese glyph forms. The Meiryo UI font being removed earlier.

The underlying issue is Han Unification - any font providing the Han unified glyphs has to choose a glyph form. Therefore deleting fonts isn't the answer, we need to provide different fonts and tell them to choose the right glyph forms for CJK languages.

We should also probably choose a default glyph form to display - likely that is simplified Chinese in honour of @lunny one of our owners.

This PR proposes that we should set up CJK language specific font-families and then set these as appropriate. It doesn't deal with the issue of documents that should contain different languages, in say markdown etc. That should be dealt with at a later point.

### CODE REVIEW

**Comment by zeripath** [comment]
Not having these fonts by default and not being aware of what the appropriate appearances are I would appreciate comments on the font stacks. @lunny, @xps2 are these stacks ok?

**Comment by lunny** [comment]
image
It seems no effective.

**Comment by zeripath** [comment]
Agh so that's the reason for the !important - sorry I removed that.

Have you set your Gitea language to Chinese? That should also change the font stack - I don't see the lang matcher in your screenshot

**Comment by zeripath** [comment]
@lunny, I've added back in the !important to override the semantic-ui css.

I think this probably isn't the correct way to do it, but it's the way we've been doing it. 

I've also added a reoverride to restore the icons.

Any chance you could retest?

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
I have set the welcome page's language as Simplified Chinese, but it seems the `{{.Language}}` return empty.
```
<html lang="">
```

**Comment by zeripath** [comment]
Heya @lunny, it's unfortunately not just a template and less change so you need to rebuild gitea to get it to put the language into the context.

I've put some more changes in to simply override the semantic css where necessary rather than use important. It's horrible but without it icons and monospace won't work properly.

**Comment by appleboy** [comment]
@zeripath Conflicts.

**Comment by zeripath** [comment]
Unfortunately any time the master CSS file changes, prs that touch it will conflict.

**Comment by lunny** [comment]
@xps2 could you confirm that this PR fixed your problem?

**Comment by xps2** [comment]
@lunny I think that it's not particularly a problem.

**Comment by zeripath** [comment]
@xps2 I don't understand your comment. Does this solve the problem of Kanji being displayed as Hanji? Or is there another problem?

Is it that you normally run your gitea in en-US but then put text in in Japanese - expecting the Kanji to remain as Kanji but they're being displayed as Hanji instead?

**Comment by xps2** [comment]
@zeripath Sorry, my bad English.
The problem was solved. I confirmed that there is no particular problem. thank you!!

**Comment by zeripath** [comment]
No worries @xps2 I thought that was what you were saying. In particular, are all the icon fonts still working?

Now, because I had to change font stacks in lot of places as I couldn't just use !important any more (it broke the icon fonts), it's worth someone with an eye for this checking that I haven't unintentionally changed the fonts in random places. 

@Silverwind as our resident UI expert do you think I've missed setting the font anywhere? Or do you have any other recommendations? (I understand you were suggesting a native font stack at some point?)

File: package-lock.json [hunk 1]
```diff
@@ -201,7 +201,6 @@
       "resolved": "https://registry.npmjs.org/combined-stream/-/combined-stream-1.0.7.tgz",
       "integrity": "sha512-brWl9y6vOB1xYPZcpZde3N9zDByXTosAeMDo4p1wzo6UMOX4vumB+TP1RZ76sfE6Md68Q0NJSrE/gbezd4Ul+w==",
       "dev": true,
-      "optional": true,
```
Comment (silverwind) [inline package-lock.json]: These changes look unrelated here. I last generated this file using npm 6.7.0, maybe a older version was used here.

File: public/less/_base.less [hunk 2]
```diff
@@ -1,11 +1,85 @@
...
+    .ui.input>input, .ui.input input,
+    .ui.statistics .statistic > .value,
+    .ui.statistic > .value,
+    .ui.statistics .statistic > .label,
+    .ui.statistic > .label,
+    .ui.steps .step .title,
+    .ui.text.container
+    {
+        font-family: "Lato", @fonts;
+    }
```
Comment (silverwind) [inline public/less/_base.less:93]: What is this abomination of a selector list trying to do? Seems quite unmaintainable to me.

Comment (zeripath) [reply public/less/_base.less:93]: That's to override semantic ui's font family

File: package-lock.json (hunk 1 above)
Comment (zeripath) [reply package-lock.json]: Ah possibly - I was confused when that appeared in the merge.

**Comment by silverwind** [comment]
Hmm I think I'd prefer a simpler approach with just two global font stacks. One for sans-serif and one for monospace. They can still be different based on language. Why not

````css
* { font: $sans-serif-fonts !important }
pre, code { font: $monspace-fonts !important }
````

This should override anything that is not `!important` (I hope semantic does not do that). For things like headings, I suggest using different font weights (or variants) instead of a different font if you really want to distinguish them, but I think non-bold headings are usually fine.

I'd personally prefer to drop Lato and switch to a system font stack because any web font is going to add a flash of invisible fonts on the first page load.

**Comment by zeripath** [comment]
With the lang() that breaks the icon fonts

**Comment by silverwind** [comment]
Right, but I guess we can `*:not(.iconclass)` or similar to exclude them.

**Comment by zeripath** [comment]
Ok, so that's not(.icon .fontawesome .octicon .octicon-meta) I think there's some others...

**Comment by zeripath** [comment]
Probably the correct option is to switch from semantic-min.css and use their less. Then we can avoid any use of !important

**Comment by silverwind** [comment]
Maybe `*:not(i::before)` might suffice. Oh I wish they would just use proper SVG icons, then we wouldn't have to deal with this abuse of fonts for icons.

**Comment by silverwind** [comment]
> use their less

I guess it depends on how their variables are set up. If they inherit from a single variable and that one is overrideable, then yes, it might work.

**Comment by zeripath** [comment]
It does have a set of variables, yes. @headerFont @pageFont. They're supposed to be put in the globals/site.variables file.

**Comment by zeripath** [comment]
Hmm thinking on, I don't think that that is going to work. Although semantic does have the variables you still can't hook in per lang font overrides at the right place...

I wonder, would sticking an inherit in the header and page font variable declarations allow the lang fonts to override?

**Comment by zeripath** [comment]
Ok, I don't mind - we can use !important and the .not (although that may be storing up trouble) or if adding inherit works we can move to less.

In terms of native vs. web fonts I've no strong feelings, however I think it would technically be a redesign. I'm not sure that Lato and the like are crucial to our brand though.

It's probably possible to reduce the risk of the flash of unstyled content though. I guess screenshots are the way to go to prove that Lato really isn't necessary.

**Comment by silverwind** [comment]
Also no strong opinion on english font. Custom font's do have a bit of value as they sometimes just look better.

Gogs has recently adopted a font called [PingFang SC](https://try.gogs.io/user/login) and I have to say I like it more than Lato. Edit: might not actually be that font rendering the english glyphs.

**Comment by zeripath** [comment]
@silverwind PingFang SC is a simplified Chinese font provided with Mac. At least on my Linux machine running Chrome I'm actually getting Microsoft YaHei providing the latin glyphs! (I'm not sure how that's happening.)

**Comment by silverwind** [comment]
Indeed, Firefox on Windows shows Microsoft YaHei as active font for me too. On macOS I get PingFang SC Regular system which seems to be a rather new font Apple introduced in 10.12. Both do seem to render rather nice-looking latin glyphs. Here's a sample from macOS:

screenshot 2019-02-20 at 22 34 07

Maybe PingFang is worth adding to the stack as the preferred macOS Font.

**Comment by zeripath** [comment]
I'm not sure where to put the `*:not(i::before)` @silverwind do you have a good example?

**Comment by silverwind** [comment]
I guess somewhere in the start of `_base.less` should be fine, position doesn't really matter for it.

**Comment by lunny** [comment]
conflicted

**Comment by CL-Jeremy** [comment]
> Gogs has recently adopted a font called PingFang SC and I have to say I like it more than Lato. Edit: might not actually be that font rendering the english glyphs.

I can confirm it is rendering English glyphs, as it is placed right at the beginning of the fallback list, a decision I cannot put up with. Gitea is definitely ahead regarding i18n!

> Maybe PingFang is worth adding to the stack as the preferred macOS Font.

Yes, please! The stacks in https://github.com/go-gitea/gitea/pull/6007/commits/d014ae5d994a0696215e728db43b874d7b492582 look thrilling indeed, but I don't get why Microsoft YaHei gets prioritized over PingFang SC. My current private settings look like this:

```
:lang(zh-Hans) {
  font-family: Lato, -apple-system, BlinkMacSystemFont, system-ui, "Segoe UI", Roboto, Helvetica, Arial, "PingFang SC", "Hiragino Sans GB", "Noto Sans CJK SC", "Source Han Sans SC", "Source Han Sans CN", "Microsoft YaHei", "Heiti SC", "SimHei", sans-serif;
}
:lang(zh-Hant) {
  font-family: Lato, -apple-system, BlinkMacSystemFont, system-ui, "Segoe UI", Roboto, Helvetica, Arial, "PingFang TC", "Hiragino Sans TC", "Noto Sans CJK TC", "Source Han Sans TC", "Source Han Sans TW", "Microsoft JhengHei", "Heiti TC", "PMingLiU", sans-serif;
}
:lang(ja) {
  font-family: Lato, -apple-system, BlinkMacSystemFont, system-ui, "Segoe UI", Roboto, Helvetica, Arial, "Hiragino Kaku Gothic ProN", "Yu Gothic", "Noto Sans CJK JP", "Source Han Sans JP", "Droid Sans Japanese", "Meiryo", "MS PGothic", sans-serif;
}
```
Some humble personal opinions:
* Regarding the localized font names: I'm not sure if they are needed for Windows, but definitely not on Mac or Linux. Plus there is no 方体/方體 installed anywhere by default: those are the suggested names by the ROC Ministry of Education and never really adopted widely; PingFang fonts are called {苹|蘋}方-{{简|簡}|繁|港}, and Heiti SC/TC 黑{体|體}-{{简|簡}|繁} respectively 
* For Hiragino, I can't speak for the tastes of people living in Japan, but the JIS 2004 standard should be complied to be at least consistent with Meiryo and Noto Sans CJK. The W3 ending should also be ditched: first there has long been a W4 variant since El Capitan, and second the boldface wouldn't be used with explicit weight reference (result: "ヒラギノ角ゴ Pro W3" -> "ヒラギノ角ゴシック ProN", or maybe even just "Hiragino Sans"?)
* Yu Gothic is available since Windows 8.1 and should be added before the free fonts. My rule of thumb: current Mac font > current Windows font > current free fonts >> deprecated OEM free fonts > deprecated OEM commercial fonts >> fallbacks
* Latin fonts almost never get used when placed after CJK fonts, so Helvetica and Arial should be placed right after Segoe UI (I respect your choice with Helvetica Neue otherwise, as I treat that font as the default iOS 7-8/OS X 10.10/HTC Sense system font and should never be used unless native Latin design is wished, so no comment on that)
* Just in case there is need for pan-Cyrillic improvements: PT Sans could be appended to the regular list (but that could be another topic)

**Comment by lafriks** [comment]
Conflict

**Comment by CL-Jeremy** [comment]
Seems obvious that such a thorough rewrite of `index.css` would result in a merge conflict... not sure if I should open another pull request or should I remain with this one, but I cloned the fork by @zeripath and reverted the two `.css` files to allow it to merge (just remember to do `make generate-stylesheets` to update the actual stylesheets afterwards), plus several fixes by myself (not sure if done right, especially the part with updating Lato, see below). I merged them into a patch: https://github.com/CL-Jeremy/gitea/commit/47f323b862b96f14db61d3562a6a6ef04affacce **(latest edit: actually fixing language chooser)** and hope that could help resolve this issue.

Remark on updating Lato: the original files (same as on Google Fonts) are significantly out of date, but I didn't touch them considering the lack of legacy iOS support with the updated files. Instead I just dropped in the new files (larger, but not really a big issue for me, at least not as serious as the inconsistencies with Cyrillic/Latin-Extended rendering). Of course this wouldn't matter if we were to ditch Lato in favor of native system fonts.

**Comment by apricote** [comment]
We are having a discussing in #6203 on how to avoid these merge conflicts for css (and potentially js) bundles.

**Comment by zeripath** [comment]
Thanks @CL-Jeremy for your changes. 

So the major negative thing that is left in this PR is the use of the specific semanticUI overrides. The main benefit of doing it that way is that you don't have `!important` with weird pre-selector tricks - which I don't know to implement - however, the cost is that there is an ugly block. (But as we move from semantic UI these can disappear, and without the `!important` we should have predictable reactions as we add other things.

**Comment by lunny** [comment]
Unfortunately, this is still unavailable for Chinese language environment. The sementic ui will override index.css. I think maybe we could move this to v1.9 @zeripath 

image

**Comment by techknowlogick** [comment]
Moving to 1.9

**Comment by CL-Jeremy** [comment]
> image

Are you sure you have used the latest code base? I believe I have cleared the last trace of these general rules in the latest commit. Probably you used my commit without `make generate-stylesheets`, while @zeripath updated again in https://github.com/go-gitea/gitea/pull/6007/commits/3034e23827cf0809ef761ace3cedd66c8088823a to include the correctly compiled stylesheets? This commit should be free from any font name in double quotes, while your screenshot still shows them.

If you may, please check the instance on https://mikeslab.dix.asia/gitea, which is built with the changes from this  PR, plus the deletion of `-webkit-font-smoothing: antialiased;` (for which I'd like to raise a separate PR some time). I am yet to find any display problem there regarding fonts.

Bildschirmfoto 2019-03-17 um 15 44 21

**Comment by zeripath** [comment]
@lunny it seems you've never had this working for you. Are you sure you've not ended up with cached templates or stylesheets?

It's worth noting it won't select the Chinese stack unless you set the language to Chinese.

**Comment by lunny** [comment]
@zeripath @CL-Jeremy It seems it's my Chrome cache problem? I tested it again on my Firefox, it's OK for Simple Chinese

**Review by lunny — APPROVED** [review]

**Comment by zeripath** [comment]
So in 1.8?
